src/workerd/io/trace-common.h

```diff
@@ -220,30 +359,192 @@ struct Log final {
+  Subrequest(rpc::Trace::Subrequest::Reader reader);
+  Subrequest(Subrequest&&) = default;
+  Subrequest& operator=(Subrequest&&) = default;
+  KJ_DISALLOW_COPY(Subrequest);
+
+  uint32_t id;
+  kj::Maybe<Info> info;
+
+  void copyTo(rpc::Trace::Subrequest::Builder builder) const;
+  Subrequest clone() const;
+};
+
+struct SubrequestOutcome final {
+  using Info = kj::OneOf<FetchResponseInfo, Tags>;
+  explicit SubrequestOutcome(
+      uint32_t id, kj::Maybe<Info> info, rpc::Trace::Span::SpanOutcome outcome);
+  SubrequestOutcome(rpc::Trace::SubrequestOutcome::Reader reader);
+  SubrequestOutcome(SubrequestOutcome&&) = default;
+  SubrequestOutcome& operator=(SubrequestOutcome&&) = default;
+  KJ_DISALLOW_COPY(SubrequestOutcome);
+
+  uint32_t id;
+  kj::Maybe<Info> info;
+  rpc::Trace::Span::SpanOutcome outcome;
+
+  void copyTo(rpc::Trace::SubrequestOutcome::Builder builder) const;
+  SubrequestOutcome clone() const;
+};
+
+struct SpanEvent final {
+  using Outcome = rpc::Trace::Span::SpanOutcome;
+  using Info = kj::OneOf<FetchResponseInfo, ActorFlushInfo, Tags>;
+  explicit SpanEvent(uint32_t id,
+      uint32_t parentId,
+      rpc::Trace::Span::SpanOutcome outcome,
+      bool transactional = false,
+      kj::Maybe<Info> maybeInfo = kj::none,
+      Tags tags = nullptr);
+  SpanEvent(rpc::Trace::Span::Reader reader);
+  SpanEvent(SpanEvent&&) = default;
+  SpanEvent& operator=(SpanEvent&&) = default;
+  KJ_DISALLOW_COPY(SpanEvent);
+
+  uint32_t id;
+  uint32_t parent;
+  rpc::Trace::Span::SpanOutcome outcome;
+  bool transactional;
+  kj::Maybe<Info> info;
+  Tags tags;
+
+  void copyTo(rpc::Trace::Span::Builder builder) const;
+  SpanEvent clone() const;
+};
+
+struct Mark final {
+  explicit Mark(kj::String name);
+  Mark(rpc::Trace::Mark::Reader reader);
+  Mark(Mark&&) = default;
+  Mark& operator=(Mark&&) = default;
+  KJ_DISALLOW_COPY(Mark);
 
-  void copyTo(rpc::Trace::Exception::Builder builder);
+  kj::String name;
+
+  void copyTo(rpc::Trace::Mark::Builder builder) const;
+  Mark clone() const;
 };
 
+struct Metric final {
+  using Key = kj::OneOf<kj::String, uint32_t>;
+  using Value = kj::OneOf<double, int64_t, uint64_t>;
+  using Type = rpc::Trace::Metric::Type;
+
+  explicit Metric(Type type, Key key, Value value);
+
+  template <IsEnum K>
+  explicit Metric(Type type, K key, Value value)
+      : Metric(type, static_cast<uint32_t>(key), kj::mv(value)) {}
+
+  Metric(rpc::Trace::Metric::Reader reader);
+  Metric(Metric&&) = default;
+  Metric& operator=(Metric&&) = default;
+  KJ_DISALLOW_COPY(Metric);
+
+  Type type;
+  Key key;
+  Value value;
+
+  bool keyMatches(kj::OneOf<kj::StringPtr, uint32_t> key);
+
+  template <IsEnum Key>
+  inline bool keyMatches(Key key) {
+    return keyMatches(static_cast<uint32_t>(key));
+  }
+
+  void copyTo(rpc::Trace::Metric::Builder builder) const;
+  Metric clone() const;
+};
+using Metrics = kj::Array<Metric>;
+
+struct Dropped final {
```

Comment:
Maybe we should change the name of the struct to be more verbose?

Reply:
Open to suggestions